pulp_container/app/registry_api.py

```diff
@@ -345,36 +349,132 @@ def get(self, request):
         return Response(data={})
 
 
-class CatalogView(ContainerRegistryApiMixin, APIView):
+class ContainerCatalogSerializer(ModelSerializer):
+    """
+    Serializer for Distributions in the _catalog endpoint of the registry.
+    """
+
+    class Meta:
+        model = models.ContainerDistribution
+        fields = ["base_path"]
+
+
+class ContainerCatalogPagination(BasePagination):
+    """
+    Pagination class to paginate repositories by names according to the registry api specification.
+    """
+
+    def paginate_queryset(self, queryset, request, view=None):
+        """
+        Analyse the pagination parameters and prepare the queryset.
+        """
+        try:
+            self.n = int(request.query_params.get("n"))
+        except Exception:
+            self.n = self.page_size
+        last = request.query_params.get("last")
+        self.url = request.build_absolute_uri()
+
+        if last:
+            queryset = queryset.filter(base_path__gt=last)
+        return queryset.order_by("base_path")[: self.n]
+
+    def get_paginated_response(self, data):
+        """
+        Prepare the paginated container _catalog response.
+        """
+        repositories_names = [repo["base_path"] for repo in data]
+        parsed_url = list(urlparse(self.url))
+        query_params = parse_qs(parsed_url[4])
+        query_params["n"] = str(self.n)
+        query_params["last"] = repositories_names[-1]
+        parsed_url[4] = urlencode(query_params)
+        url = urlunparse(parsed_url)
+        log.info(url)
+        headers = {
+            "Link": f'<{url}>; rel="next"',
+        }
+        return Response(headers=headers, data={"repositories": list(repositories_names)})
```

Comment:
repo_name is already a list

Reply:
i see a problem with this approach - the link header should be present in the response only when it indicates that not all the results have been displayed.
This way we will always have link in the header which is wrong.

Reply:
```
(pulp) [vagrant@pulp3-source-fedora32 ~]$ http GET :24817/v2/test/this/tags/list
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, HEAD, OPTIONS
Connection: close
Content-Length: 40
Content-Type: application/json
Correlation-ID: 39902f76fc6a4b2696d31040ceeba9ec
Date: Wed, 17 Feb 2021 17:33:41 GMT
Docker-Distribution-Api-Version: registry/2.0
Link: <http://localhost:24817/v2/test/this/tags/list?n=16&last=mytag1.8>; rel="next"
Server: gunicorn/20.0.4
Vary: Accept
X-Frame-Options: SAMEORIGIN

{
    "name": "test/this",
    "tags": [
        "mytag1.8"
    ]
}

```

Reply:
however i agree on the approach to paginate the response even if the client did not sent request with pagination.

Reply:
Not sure if we can get away with this. If you follow the next link here (provided i fixed the empty list problem) you'll get an empty list, and you stop iterating.
But i will try to find a solution without this extra call anyway.

Reply:
Now, with the empty list, there will be no next link anymore. But as long as we return something, i do not see, where i can get the information that there would be more.

Reply:
one more thing i would suggest is to return in the link header a relative URI instead of absolute. According to RFC both are support though